jsparagus/parse_table.py

```diff
@@ -1314,6 +1322,138 @@ def remove_all_unreachable_state(self, verbose: bool, progress: bool) -> None:
+                    # have a single path for each reduced path. If the next
+                    # state contains an action, then we stop here.
+                    iter_aps = reduced_aps
+                    next_is_action = self.states[iter_aps.state].epsilon != []
+                    has_replay = iter_aps.replay != []
+                    assert next_is_action is False and has_replay is True
+                    while (not next_is_action) and has_replay:
+                        # print("Step {}:\n{}".format(len(iter_aps.history),
+                        #                             iter_aps.string(name="\titer_aps")))
+                        next_aps = list(iter_aps.shift_next(self))
+                        if len(next_aps) == 0:
+                            # Note, this might happen as we are adding
+                            # lookahead tokens from any successor, we might not
+                            # always have a way to replay all tokens, in such
+                            # case an error should be produced, but in the mean
+                            # time, let's use the shift function as usual.
+                            break
+                        assert len(next_aps) == 1
+                        iter_aps = next_aps[0]
+                        next_is_action = self.states[iter_aps.state].epsilon != []
+                        has_replay = iter_aps.replay != []
+                    # print("End at {}:\n{}".format(len(iter_aps.history),
+                    #                               iter_aps.string(name="\titer_aps")))
+                    replay_list = list(map(lambda e: e.src, iter_aps.shift))
+                    assert len(replay_list) >= 2
+                    replay_term = Replay(replay_list[1:])
+                    states_by_replay_term[replay_term].append(replay_list[0])
+
+                # Create FilterStates actions.
+                filter_by_replay_term = {
+                    replay_term: FilterStates(states)
+                    for replay_term, states in states_by_replay_term.items()
+                }
+
+                # Convert the Reduce action to an Unwind action.
+                reduce_term, _ = next(iter(s.epsilon))
+                if isinstance(reduce_term, Reduce):
+                    unwind_term: Action = reduce_term.unwind
+                else:
+                    assert isinstance(reduce_term, Seq)
+                    assert isinstance(reduce_term.actions[-1], Reduce)
+                    unwind_term = Seq(list(reduce_term.actions[:-1]) + [reduce_term.actions[-1].unwind])
+
+                # Remove the old Reduce edge if still present.
+                # print("Before:\n{}".format(reduce_state.stable_str(self.states)))
+                self.remove_edge(reduce_state, reduce_term, maybe_unreachable_set)
+
+                # Add Unwind action.
+                # print("After:\n")
+                locations = OrderedFrozenSet(reduce_state.locations)
```

Comment:
`reduce_state.locations` is already an `OrderedFrozenSet`, so calling `OrderedFrozenSet` on it makes an unnecessary copy.